mathics/core/expression.py

```diff
@@ -297,7 +299,12 @@ def _build_elements(
 
         last_converted_elt = None
         for element in elements:
-            converted_elt = conversion_fn(element)
+            # FIXME: we should have isinstance(element(BaseElement, Builtin))
```

Comment:
Actually, what should not be converted are InstanceableBuiltins. And I agree: these elements in the future should be a subclass of BaseElement.

Reply:
Ok. Will change that later. 

I also wonder if this shouldn't be done instead in from_python.... benchmarking will tell us though.

Reply:
Perhaps Instanceable Builtins would be better named BuiltinElement to replace something about the mechanism (is some sort of thing that we get an instance out of) with something about its intended use (acts as an element in an Expression).

Reply:
Yep, I agree.

Reply:
> Ok. I will change that later.
> 
> I also wonder if this shouldn't be done instead in from_python.... benchmarking will tell us though.

Actually, it is already done here:
https://github.com/Mathics3/mathics-core/blob/1106ff85ac40e5d359232a4e168118d516f73691/mathics/core/atoms.py#L1076

The problem is that making function calls in Python is onerous. This is why, for example, I would prefer to avoid the call at all instead of calling something like `identity_fn`.
Also, having an optional parameter adds some overhead:
```
li = [i for i in range(10)]        

def identity_fn(x):
    return x

def  f0(*u):
    return None


def  f1(*u, g=identity_fn):
    return None

def  f2(*u, **kwargs):
    g = kwargs.get("g", identity_fn)
    return None

%timeit f0(li)  
%timeit f1(li)
%timeit f2(li)
```
gives
```
50.4 ns ± 1.04 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
58.7 ns ± 4.85 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
79 ns ± 2.55 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
This is why I think that we do not gain too much in most cases by doing the conversion inside `_build_elements`. Mostly because in the most common calls (which come from the calls inside `rewrite_apply_eval_step`) are that the arguments are already Mathics Elements.

Reply:
@rocky, the point here was not to decide which of these functions are better defined but to have an idea of the overall factor introduced by adding optional parameters in a very frequently called function.   If you call 1000 times a 50ns function, the 50ns becomes 50 in microseconds. 
On the other hand, if you are not worried about the cost of a call, you shouldn't worry either about the number of calls to `from_python`: most of the time, this function does nothing, because, on its first line, it checks if the conversion is needed.